poc/flp_pine.py

```diff
@@ -195,6 +174,170 @@ def eval(self,
+            # and the field modulus, and return the result as negative.
+            decoded = -(field_elem.MODULUS - decoded)
+        # Divide by 2^num_frac_bits and we will get a float back.
+        decoded_float = decoded / (2 ** self.num_frac_bits)
+        return decoded_float
+
+
+class PineValidSquaredNormEqualityCheck(
+    PineValidBase[F]
+):
+    """PINE validation circuit that computes the squared L2-norm result
+    and checks that it matches the value claimed by the Client.
+        * This circtuit does not use joint randomness.
+        * The result is only valid if the other circtuit 
+        `PineValid` is also valid.
+    """
+
+    def __init__(self,
+                 field: type[F],
+                 l2_norm_bound: int,
+                 num_frac_bits: int,
+                 dimension: int,
+                 chunk_length: int,
+                 alpha: float = ALPHA,
+                 num_wr_checks: int = NUM_WR_CHECKS,
+                 num_wr_successes: int = NUM_WR_SUCCESSES):
+        super().__init__(field,
+                         l2_norm_bound,
+                         num_frac_bits,
+                         dimension,
+                         chunk_length,
+                         alpha,
+                         num_wr_checks,
+                         num_wr_successes)
+        self.JOINT_RAND_LEN = 0
+        self.GADGET_CALLS = [chunk_count(self.chunk_length, dimension)]
+
+    def eval(self,
+             meas: list[F],
+             joint_rand: list[F],
+             _num_shares: int) -> list[F]:
+
+        self.check_valid_eval(meas, joint_rand)
+        encoded_gradient, _bit_checked, _wr_check_results, sq_norm_v_bits, \
+            _sq_norm_u_bits, _wr_check_v_bits, _wr_check_g = self.unpack_encoded_measurement(
+                meas)
+        result = self.eval_norm_equality_check(
+            encoded_gradient, sq_norm_v_bits)
+        return [result]
+
+    def eval_norm_equality_check(self,
+                                 encoded_gradient: list[F],
+                                 sq_norm_v_bits: list[F]) -> F:
+        """Check that the computed squared L2-norm result matches the value 
+        claimed by the Client. """
+        # Compute the squared norm.
+        mul_inputs = []
+        for val in encoded_gradient:
+            mul_inputs += [val, val]
+        computed_sq_norm = self.parallel_sum(mul_inputs)
+        sq_norm_v = self.field.decode_from_bit_vector(sq_norm_v_bits)
+
+        return sq_norm_v - computed_sq_norm
+
+    def test_vec_set_type_param(self, test_vec):
+        test_vec["chunk_length_norm_equality"] = self.chunk_length
+        return ["chunk_length_norm_equality"]
+
+
+class PineValid(
+    PineValidBase[F]
+):
+    """
+    Validity circuit for all checks for PINE except the squared norm quality 
+    check. The result is only valid if the circtuit 
+    `PineValidSquaredNormEqualityCheck` is also valid.
```

Comment:
spelling and trailing whitespace
```suggestion
    Validity circuit for all checks for PINE except the squared norm quality
    check. The result is only valid if the circuit
    `PineValidSquaredNormEqualityCheck` is also valid.
```